semgrep/semgrep/join_rule.py

```diff
@@ -0,0 +1,404 @@
+    targets: List[Path],
+) -> Tuple[List[RuleMatch], List[SemgrepError]]:
+    join_contents = join_rule.get("join", {})
+    semgrep_config_strings = [ref.get("rule") for ref in join_contents.get("refs", [])]
+    config_map = create_config_map(semgrep_config_strings)
+
+    join_rule_refs: List[Ref] = [
+        Ref(
+            id=config_map[ref.get("rule")].id,
+            renames={
+                rename.get("from"): rename.get("to")
+                for rename in ref.get("renames", [])
+            },
+            alias=ref.get("as"),
+        )
+        for ref in join_contents.get("refs", [])
+    ]
+    refs_lookup = {ref.id: ref for ref in join_rule_refs}
+    alias_lookup = {ref.alias: ref.id for ref in join_rule_refs}
+
+    try:
+        conditions = [
+            Condition.parse(condition_string)
+            for condition_string in join_contents.get("on", [])
+        ]
+    except InvalidConditionError as e:
+        return [], [e]
+
+    # Run Semgrep
+    with tempfile.NamedTemporaryFile() as rule_path:
+        yaml.dump({"rules": [rule.raw for rule in config_map.values()]}, rule_path)
+        rule_path.flush()
+        output = semgrep.semgrep_main.invoke_semgrep(
+            config=Path(rule_path.name),
+            targets=targets,
+            no_rewrite_rule_ids=True,
+            optimizations="all",
+        )
+
+    assert isinstance(output, dict)  # placate mypy
+
+    results = output.get("results", [])
+    errors = output.get("errors", [])
+
+    parsed_errors = []
+    for error_dict in errors:
+        try:
+            del error_dict["code"]
+            del error_dict["level"]
+            errortype = error_dict.get("type")
+            del error_dict["type"]
+            parsed_errors.append(
+                ERROR_MAP[error_dict.get(errortype)].from_dict(error_dict)
+            )
+        except KeyError:
+            logger.warning(
+                f"Could not reconstitute Semgrep error: {error_dict}.\nSkipping processing of error"
+            )
+            continue
+
+    # Small optimization: if there are no results for rules that
+    # are used in a condition, there's no sense in continuing.
+    collection_set_unaliased = {
+        alias_lookup[collection]
+        for collection in create_collection_set_from_conditions(conditions)
+    }
+    rule_ids = set(result.get("check_id") for result in results)
+    if collection_set_unaliased - rule_ids:
+        logger.debug(
+            f"No results for {collection_set_unaliased - rule_ids} in join rule '{join_rule.get('id')}'."
+        )
+        return [], parsed_errors
+
+    # Rename metavariables with user-defined renames.
+    rename_metavars_in_place(results, refs_lookup)
+
+    # Create a model map. This allows dynamically creating DB tables based
+    # on Semgrep's results. There is one table for each rule ID.
+    model_map = create_model_map(results)
+    db.connect()
+    db.create_tables(model_map.values())
```

Comment:
Should we be deleting the database on exit/error? or are in memory dbs automatically deleted on finish?

Reply:
I assumed they were deleted when the program closed, but I can explicitly `.close()` which definitely removes the in-memory database